src/order/filter/filter_product.lean

```diff
@@ -0,0 +1,223 @@
+  right_distrib := λ x y z, quotient.induction_on₃' x y z 
+    (λ x y z, quotient.sound' (by simp only [right_distrib]; apply (setoid.iseqv _).1)), 
+  ..filter_product.has_add β φ,
+  ..filter_product.has_mul β φ }
+
+instance [mul_zero_class β] : mul_zero_class β* :=
+{ zero_mul := λ x, quotient.induction_on' x 
+    (λ a, quotient.sound' (by simp only [zero_mul]; apply (setoid.iseqv _).1)),
+  mul_zero := λ x, quotient.induction_on' x 
+    (λ a, quotient.sound' (by simp only [mul_zero]; apply (setoid.iseqv _).1)), 
+  ..filter_product.has_mul β φ,
+  ..filter_product.has_zero β φ }
+
+instance [semiring β] : semiring β* := 
+{ ..filter_product.add_comm_monoid β φ, 
+  ..filter_product.monoid β φ, 
+  ..filter_product.distrib β φ, 
+  ..filter_product.mul_zero_class β φ }
+
+instance [ring β] : ring β* := 
+{ ..filter_product.add_comm_group β φ, 
+  ..filter_product.monoid β φ, 
+  ..filter_product.distrib β φ }
+
+instance [comm_semiring β] : comm_semiring β* := 
+{ ..filter_product.semiring β φ, 
+  ..filter_product.comm_monoid β φ }
+
+instance [comm_ring β] : comm_ring β* := 
+{ ..filter_product.ring β φ, 
+  ..filter_product.comm_semigroup β φ }
+
+instance [zero_ne_one_class β] (U : is_ultrafilter φ) : zero_ne_one_class β* :=
+{ zero_ne_one := by
+  { intro c, have c' : _ := quotient.exact' c, 
+    change _ ∈ _ at c', 
+    simp only [set.set_of_false, zero_ne_one, empty_in_sets_eq_bot] at c', 
+    apply U.1 c' },
+  ..filter_product.has_zero β φ, 
+  ..filter_product.has_one β φ }
+
+instance [division_ring β] (U : is_ultrafilter φ) : division_ring β* :=
+{ mul_inv_cancel := λ x, quotient.induction_on' x $ λ a hx, quotient.sound' $
+    have hx1 : _ := (not_imp_not.mpr quotient.eq'.mpr) hx,
+    have hx2 : _ := (ultrafilter_iff_compl_mem_iff_not_mem.mp U _).mpr hx1,
+    have h : {n : α | ¬a n = 0} ⊆ {n : α | a n * (a n)⁻¹ = 1} := 
+      by rw [set.set_of_subset_set_of]; exact λ n, division_ring.mul_inv_cancel,
+    mem_sets_of_superset hx2 h,
+  inv_mul_cancel := λ x, quotient.induction_on' x $ λ a hx, quotient.sound' $
+    have hx1 : _ := (not_imp_not.mpr quotient.eq'.mpr) hx,
+    have hx2 : _ := (ultrafilter_iff_compl_mem_iff_not_mem.mp U _).mpr hx1,
+    have h : {n : α | ¬a n = 0} ⊆ {n : α | (a n)⁻¹ * a n = 1} := 
+      by rw [set.set_of_subset_set_of]; exact λ n, division_ring.inv_mul_cancel,
+    mem_sets_of_superset hx2 h, 
+  ..filter_product.ring β φ, 
+  ..filter_product.has_inv β φ, 
+  ..filter_product.zero_ne_one_class β φ U }
+
+instance [field β] (U : is_ultrafilter φ) : field β* :=
+{ ..filter_product.comm_ring β φ, 
+  ..filter_product.division_ring β φ U }
+
+noncomputable instance [discrete_field β] (U : is_ultrafilter φ) : discrete_field β* :=
+{ inv_zero := quotient.sound' $ by show _ ∈ _;
+    simp only [inv_zero, eq_self_iff_true, (set.univ_def).symm, univ_sets],
+  has_decidable_eq := by apply_instance, 
+  ..filter_product.field β φ U }
+
+end filter_product
```

Comment:
also order structures... I think linear order is the first place you really need the ultrafilter property